File: src/test/java/com/raychenon/leetcode/tree/SymmetricTreeTest.java
package com.raychenon.leetcode.tree;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static junit.framework.TestCase.assertEquals;

/**
 * User: raychenon
 * Date: 14/9/2020
 * https://leetcode.com/problems/symmetric-tree
 */
public class SymmetricTreeTest {

    private static SymmetricTree instance;

    @BeforeClass
    public static void setUp() {
        instance = new SymmetricTree();
    }

    @Test
    public void nullNodeTest() {
        TreeNode root = null;
        assertSymetricTree(true, root);
    }


    /**
     * ---- 1
     * --- / \
     * -- 2    2
     * - / \  / \
     * 3  4  4  3
     */
    @Test
    public void example1Test() {
        TreeNode root = TreeUtil.createTree(List.of(1, 2, 2, 3, 4, 4, 3));
        assertSymetricTree(true, root);
    }

    /**
     * ----1
     * -- / \
     * ---2  2
     * ----\  \
     * -----3  3
     */
    @Test
    public void example2Test() {
        TreeNode root = TreeUtil.createTree(Arrays.asList(1, 2, 2, null, 3, null, 3));
        assertSymetricTree(false, root);
    }

    private void assertSymetricTree(boolean isSymmetric, TreeNode root) {
        assertEquals(isSymmetric, instance.isSymmetricRec(root));
        assertEquals(isSymmetric, instance.isSymmetricIter(root));
    }
}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/tree/SymmetricTreeTest.kt
package com.raychenon.kotlin.leetcode.tree

import com.raychenon.leetcode.tree.TreeNode
import com.raychenon.leetcode.tree.TreeUtil
import junit.framework.TestCase
import org.junit.Test
import java.util.*
import java.util.List

/**
 * User: raychenon
 * Date: 14/9/2020
 * https://leetcode.com/problems/symmetric-tree/
 */
class SymmetricTreeTest {

    @Test
    fun nullNodeTest() {
        val root: TreeNode? = null
        assertSymetricTree(true, root)
    }


    /**
     * ---- 1
     * --- / \
     * -- 2    2
     * - / \  / \
     * 3  4  4  3
     */
    @Test
    fun example1Test() {
        val root = TreeUtil.createTree(List.of(1, 2, 2, 3, 4, 4, 3))
        assertSymetricTree(true, root)
    }

    /**
     * ----1
     * -- / \
     * ---2  2
     * ----\  \
     * -----3  3
     */
    @Test
    fun example2Test() {
        val root = TreeUtil.createTree(Arrays.asList(1, 2, 2, null, 3, null, 3))
        assertSymetricTree(false, root)
    }

    private fun assertSymetricTree(isSymmetric: Boolean, root: TreeNode?) {
        TestCase.assertEquals(isSymmetric, SymmetricTree.isSymmetricRec(root))
        //TestCase.assertEquals(isSymmetric, SymmetricTree.isSymmetricIter(root))
    }
}